VerilatedVarProps::totalSize was missing the first unpacked dim

teqdruid opened the pull request:
Throwing off the calculation. This looks like a simple off-by-one error.

The net effect is when running 'svSizeOfArray' from a DPI function, said calculation does not include the first unpacked dimension of the array. For instance, if I have a 'byte Data [3:0]', it returns 1 when it should return 4.


We appreciate your contributing to Verilator.  If this is your first commit, please add your name to docs/CONTRIBUTORS, and read our contributing guidelines in docs/CONTRIBUTING.adoc.


This is a one-character change... Is including my name in the contributors list appropriate?

Closes #2297

wsnyder commented:
Thanks for your first pull request.

Yes, please update CONTRIBUTORS, because you deserve it, and because we have a test that fails otherwise.

Please update t_dpi_open_c.cpp to test this case.  There's an interesting comment there with that function commented out that not all simulators support this call, but maybe that is no longer the case.  If possible please run this test on another simulator if you can to make sure the results match.

teqdruid commented:
Done.

Unfortunately, the only simulation to which I have access is Questa. It doesn't seem that test has support for Questa, but I commented out the proper lines to run the test, which seemed to work. I would have added that support, but I couldn't figure out the macro which Quartus sets during DPI compilation. (Or do you set NC/VCS?)

Anecdotally, we use this DPI call internally (at Microsoft) with Questa and it generally works as one would expect. I'm also told that the same DPI library works on other simulators (and its functionality relies on svSizeOfArray), but I don't recall which ones were tested.

teqdruid commented:
Also, it looks like the check above svGetArrayPtr is just a non-crash check. I had assumed that would be fine for svSizeOfArray, but I'll add a result check in my next push.

gezalore commented:
You can use the `MS` macro (ModelSim) for Questa, see here for an example:
https://github.com/verilator/verilator/blob/9fd4541069c019251a864d42fa0d1fedf8db53b0/test_regress/t/t_dpi_arg_output_type.cpp#L42-L46

Note this is set by the test driver and not Questa itself:
https://github.com/verilator/verilator/blob/9fd4541069c019251a864d42fa0d1fedf8db53b0/test_regress/driver.pl#L605

gezalore commented:
Just in case you haven't spotted it yet, you should be bale to run the test with Questa using the test driver by running `cd test_regress; ./t/t_dpi_open.pl --ms`, which will pick up vlib and vlog from your PATH.

teqdruid commented:
I've added the value correctness checks. I also added Questa/ModelSim support to that file -- I found driver.pl where you set VCS/NC/MS. I couldn't use it since I don't have easy access to Questa on Linux and couldn't get your perl script working on Windows.

Took me about an hour to figure out why not all arrays are "DpiCLayout" -- I originally had the checks in the reg[] array check location!

teqdruid commented:
Also, I have originally #ifdef'd out a number of tests which fail in Questa. I've since removed that to let the tests fail since I'm not sure why they fail.

wsnyder commented:
>Should I remove the macro stuff at the top to let this file claim not to recognize MS? Not sure how your non-Verilator tests work.

If a test passes on Questa please leave it ifdef'ed in.  If you are talking about other tests separate from this bug you want to #ifdef in Questa, please make another pull request.

BTW this still passes on VCS which is great. Do you think this is good to go?

teqdruid commented:
This originally error'd out on compile with the macro error indicating that it didn't recognize Questa. It's not my additional code that fails. I've made this test run in Questa to test my changes but other checks fail in Questa for unrelated reasons.

I'll disable Questa support -- put it back to where it was originally. Should I file a bug to investigate this Questa failure at some point in the future?

teqdruid commented:
I think it's good to go

wsnyder commented:
>Should I file a bug to investigate this Questa failure at some point in the future?

If you want to do the work feel free to make a pull request, but probably not worth having a bug stick around on it.

Once passes I'll merge
